File: src/Systems/Objective-C/state.m
#import <Foundation/Foundation.h>

@class Turnstile;

@protocol GateState <NSObject>
- (NSString *)name;
- (NSString *)coin:(Turnstile *)turnstile;
- (NSString *)push:(Turnstile *)turnstile;
@end

@interface LockedState : NSObject <GateState>
@end

@interface UnlockedState : NSObject <GateState>
@end

@interface Turnstile : NSObject {
    id<GateState> _state;
}
- (instancetype)init;
- (NSString *)stateName;
- (NSString *)coin;
- (NSString *)push;
- (void)transitionTo:(id<GateState>)state;
@end

@implementation Turnstile
- (instancetype)init {
    if ((self = [super init])) {
        _state = [LockedState new];
    }
    return self;
}

- (NSString *)stateName { return [_state name]; }
- (NSString *)coin { return [_state coin:self]; }
- (NSString *)push { return [_state push:self]; }
- (void)transitionTo:(id<GateState>)state { _state = state; }
@end

@implementation LockedState
- (NSString *)name { return @"locked"; }
- (NSString *)coin:(Turnstile *)turnstile {
    [turnstile transitionTo:[UnlockedState new]];
    return @"unlocked";
}
- (NSString *)push:(Turnstile *)turnstile {
    (void)turnstile;
    return @"blocked";
}
@end

@implementation UnlockedState
- (NSString *)name { return @"unlocked"; }
- (NSString *)coin:(Turnstile *)turnstile {
    (void)turnstile;
    return @"coin-returned";
}
- (NSString *)push:(Turnstile *)turnstile {
    [turnstile transitionTo:[LockedState new]];
    return @"passed";
}
@end

static void require(BOOL condition, NSString *message) {
    if (!condition) {
        NSLog(@"State contract failed: %@", message);
        abort();
    }
}

int main(void) {
    @autoreleasepool {
        Turnstile *turnstile = [Turnstile new];

        require([[turnstile stateName] isEqualToString:@"locked"], @"initial state");
        require([[turnstile push] isEqualToString:@"blocked"], @"invalid push result");
        require([[turnstile stateName] isEqualToString:@"locked"], @"invalid push keeps state");
        require([[turnstile coin] isEqualToString:@"unlocked"], @"coin unlocks");
        require([[turnstile stateName] isEqualToString:@"unlocked"], @"unlocked state");
        require([[turnstile coin] isEqualToString:@"coin-returned"], @"duplicate coin result");
        require([[turnstile stateName] isEqualToString:@"unlocked"], @"duplicate coin keeps state");
        require([[turnstile push] isEqualToString:@"passed"], @"push passes");
        require([[turnstile stateName] isEqualToString:@"locked"], @"push locks again");

        puts("objective-c-state: passed");
    }
    return 0;
}

File: src/Systems/Go/state.go
package main

import "fmt"

type gateState interface {
	name() string
	coin(*turnstile) string
	push(*turnstile) string
}

type lockedState struct{}
type unlockedState struct{}

type turnstile struct {
	state gateState
}

func newTurnstile() *turnstile {
	return &turnstile{state: lockedState{}}
}

func (lockedState) name() string { return "locked" }
func (lockedState) coin(t *turnstile) string {
	t.state = unlockedState{}
	return "unlocked"
}
func (lockedState) push(*turnstile) string { return "blocked" }

func (unlockedState) name() string           { return "unlocked" }
func (unlockedState) coin(*turnstile) string { return "coin-returned" }
func (unlockedState) push(t *turnstile) string {
	t.state = lockedState{}
	return "passed"
}

func require(ok bool) {
	if !ok {
		panic("Go State contract failed")
	}
}

func main() {
	gate := newTurnstile()
	require(gate.state.name() == "locked")

	require(gate.state.push(gate) == "blocked")
	require(gate.state.name() == "locked")

	require(gate.state.coin(gate) == "unlocked")
	require(gate.state.name() == "unlocked")

	require(gate.state.coin(gate) == "coin-returned")
	require(gate.state.name() == "unlocked")

	require(gate.state.push(gate) == "passed")
	require(gate.state.name() == "locked")

	fmt.Println("go-state: passed")
}
